ticdc/ticdc-canal-json.md

```diff
@@ -0,0 +1,455 @@
+| pkNames   | Array  | The names of all the columns that make up the primary key                |
+| isDdl     | Bool   | Whether the message is a DDL event                                       |
+| type      | String | Event types defined by Canal-JSON                                        |
+| es        | Number | 13-bit (millisecond) timestamp when the event that generated the message happened     |
+| ts        | Number | 13-bit (millisecond) timestamp when TiCDC generated the message                       |
+| sql       | String | When isDdl is `true`, records the corresponding DDL statement                           |
+| sqlType   | Object | When isDdl is `false`, records how the data type of each column is represented in Java  |
+| mysqlType | object | When isDdl is `false`, records how the data type of each column is represented in MySQL |
+| data      | Object | When isDdl is `false`, records the name of each column and its data value       |
+| old       | Object | Only if the message is generated by an update Event, records the name of each column and the data value before the update |
+| _tidb     | Object | TiDB extension field. It exists only if you set `enable-tidb-extension` to `true`. The value of `commitTs` is the TSO of the transaction that caused the row to change. |
+
+### DML Event
+
+TiCDC encodes a row of DML data change event as follows:
+
+```json
+{
+    "id": 0,
+    "database": "test",
+    "table": "tp_int",
+    "pkNames": [
+        "id"
+    ],
+    "isDdl": false,
+    "type": "INSERT",
+    "es": 1639633141221,
+    "ts": 1639633142960,
+    "sql": "",
+    "sqlType": {
+        "c_bigint": -5,
+        "c_int": 4,
+        "c_mediumint": 4,
+        "c_smallint": 5,
+        "c_tinyint": -6,
+        "id": 4
+    },
+    "mysqlType": {
+        "c_bigint": "bigint",
+        "c_int": "int",
+        "c_mediumint": "mediumint",
+        "c_smallint": "smallint",
+        "c_tinyint": "tinyint",
+        "id": "int"
+    },
+    "data": [
+        {
+            "c_bigint": "9223372036854775807",
+            "c_int": "2147483647",
+            "c_mediumint": "8388607",
+            "c_smallint": "32767",
+            "c_tinyint": "127",
+            "id": "2"
+        }
+    ],
+    "old": null,
+    "_tidb": {     // TiDB extension field
+        "commitTs": 163963314122145239
+    }
+}
+```
+
+### WATERMARK Event
+
+TiCDC sends a WATERMARK Event only when you set `enable-tidb-extension` to `true`. The value of the `type` field is `TIDB_WATERMARK`. The Event contains the `_tidb` field that contains only `watermarkTs`. The value of `watermarkTs` is the TSO recorded when the Event is sent.
+
+When you receive an Event of this type, all Events with `commitTs` less than `watermarkTs` have been sent. Because TiCDC provides the "At Least Once" semantics, data might be sent repeatedly. If a subsequent Event with `commitTs` less than `watermarkTs` is received, you can safely ignore it.
```

Comment:
```suggestion
When you receive an Event of this type, all Events with `commitTs` less than `watermarkTs` have been sent. Because TiCDC provides the "At Least Once" semantics, data might be sent repeatedly. If a subsequent Event with `commitTs` less than `watermarkTs` is received, you can safely ignore this Event.
```